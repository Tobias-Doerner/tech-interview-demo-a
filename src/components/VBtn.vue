<template>
  <button
    :disabled="disabled"
    :class="classObject"
    class="base flex items-center"
    @click="$emit('click', $event)"
  >
    <slot></slot>
    <div class="spinner" v-if="loading" />
  </button>
</template>

<script>
export default {
  props: {
    disabled: {
      type: Boolean,
      default: false
    },
    loading: {
      type: Boolean,
      default: false
    },
    primary: {
      type: Boolean,
      default: false
    },
    secondary: {
      type: Boolean,
      default: false
    }
  },
  computed: {
    classObject() {
      if (this.primary) return { primary: true }
      if (this.secondary) return { secondary: true }
      return { primary: true }
    }
  }
}
</script>

<style scoped>
.base {
  @apply font-bold px-2 py-1 rounded-sm text-white disabled:opacity-50;
}

.primary {
  @apply bg-blue-500 active:bg-blue-800 hover:bg-blue-700;
}

.secondary {
  @apply bg-gray-500 active:bg-gray-800 hover:bg-gray-700;
}

.spinner {
  @apply mx-1 animate-spin rounded-full h-5 w-5 border-b-4 border-white;
}
</style>
