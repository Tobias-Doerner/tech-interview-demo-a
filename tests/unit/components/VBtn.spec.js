import { mount, shallowMount } from '@vue/test-utils'
import VBtn from '@/components/VBtn'

describe('VBtn', () => {
  describe('computed properties', () => {
    test('classObject should set primary', () => {
      const wrapper = shallowMount(VBtn, {
        propsData: {
          primary: true,
          secondary: false
        }
      })
      expect(wrapper.vm.classObject).toEqual({ primary: true })
    })

    test('classObject should set primary', () => {
      const wrapper = shallowMount(VBtn, {
        propsData: {
          primary: true,
          secondary: true
        }
      })
      expect(wrapper.vm.classObject).toEqual({ primary: true })
    })

    test('classObject should set secondary', () => {
      const wrapper = shallowMount(VBtn, {
        propsData: {
          primary: false,
          secondary: true
        }
      })
      expect(wrapper.vm.classObject).toEqual({ secondary: true })
    })

    test('classObject should set primary as default', () => {
      const wrapper = shallowMount(VBtn)
      expect(wrapper.vm.classObject).toEqual({ primary: true })
    })
  })

  describe('events', () => {
    test('button should emit click event', async () => {
      const wrapper = shallowMount(VBtn)
      await wrapper.find('button').trigger('click')
      expect(wrapper.emitted().click).toBeTruthy()
    })
  })

  describe('snapshot tests', () => {
    test('should render button with loading spinner correctly', () => {
      const wrapper = mount(VBtn, {
        propsData: {
          loading: true
        },
        slots: {
          default: 'Submit'
        }
      })
      expect(wrapper.element).toMatchSnapshot()
    })

    test('should render button without loading spinner correctly', () => {
      const wrapper = mount(VBtn, {
        slots: {
          default: 'Submit'
        }
      })
      expect(wrapper.element).toMatchSnapshot()
    })
  })
})
