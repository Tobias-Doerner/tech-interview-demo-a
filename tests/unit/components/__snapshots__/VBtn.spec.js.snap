// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VBtn snapshot tests should render button with loading spinner correctly 1`] = `
<button
  class="base flex items-center primary"
>
  Submit 
  <div
    class="spinner"
  />
</button>
`;

exports[`VBtn snapshot tests should render button without loading spinner correctly 1`] = `
<button
  class="base flex items-center primary"
>
  Submit 
  <!---->
</button>
`;
